File: ra-idp-go/internal/authentication/usecases/change_password.go
package usecases

import (
	"context"
	"errors"
	"slices"
	"time"

	authports "ra-idp-go/internal/authentication/ports"
	oauthports "ra-idp-go/internal/oauth2/ports"
	"ra-idp-go/internal/spec"
)

var (
	ErrCurrentPasswordMismatch = errors.New("current password does not match")
	ErrPasswordReused          = errors.New("new password matches a recent password")
	ErrUserNotFound            = errors.New("user not found")
)

type ChangePasswordInput struct {
	Sub             string
	CurrentPassword string
	NewPassword     string
	Now             time.Time
}

type ChangePasswordDeps struct {
	UserRepo            oauthports.UserRepository
	PasswordHasher      authports.PasswordHasher
	PasswordHistoryRepo authports.PasswordHistoryRepository
	Emit                func(spec.DomainEvent)
	HistoryDepth        int                    // Deprecated: use Policy 指定。後方互換のためのフォールバック。
	Policy              PasswordPolicySnapshot // テナント解決済みのしきい値。ゼロ値は global default。
}

func ChangePassword(ctx context.Context, deps ChangePasswordDeps, in ChangePasswordInput) (*spec.User, error) {
	user, err := deps.UserRepo.FindBySub(ctx, in.Sub)
	if err != nil {
		return nil, err
	}
	if user == nil {
		return nil, ErrUserNotFound
	}

	ok, err := deps.PasswordHasher.Verify(in.CurrentPassword, user.PasswordHash)
	if err != nil {
		return nil, err
	}
	if !ok {
		return nil, ErrCurrentPasswordMismatch
	}

	snap := resolveSnapshot(deps.Policy, deps.HistoryDepth)
	result := ValidatePasswordWith(in.NewPassword, snap)
	if !result.OK {
		return nil, &PasswordPolicyError{Violations: result.Violations}
	}

	depth := snap.HistoryDepth
	recent, err := deps.PasswordHistoryRepo.Recent(ctx, user.Sub, depth)
	if err != nil {
		return nil, err
	}
	for _, entry := range recent {
		matched, err := deps.PasswordHasher.Verify(in.NewPassword, entry.Encoded)
		if err != nil {
			return nil, err
		}
		if matched {
			return nil, ErrPasswordReused
		}
	}
	matched, err := deps.PasswordHasher.Verify(in.NewPassword, user.PasswordHash)
	if err != nil {
		return nil, err
	}
	if matched {
		return nil, ErrPasswordReused
	}

	now := in.Now.UTC()
	if now.IsZero() {
		now = time.Now().UTC()
	}
	encoded, err := deps.PasswordHasher.Hash(in.NewPassword)
	if err != nil {
		return nil, err
	}

	updated := *user
	updated.PasswordHash = encoded
	updated.UpdatedAt = now
	updated.Lifecycle.PasswordChangedAt = &now
	// 本人がパスワードを変更したので update_password の強制アクションは自動解除する。
	clearedUpdatePassword := slices.Contains(updated.Lifecycle.RequiredActions, spec.RequiredActionUpdatePassword)
	if clearedUpdatePassword {
		updated.Lifecycle.RequiredActions = removeRequiredAction(
			updated.Lifecycle.RequiredActions, spec.RequiredActionUpdatePassword,
		)
	}
	if err := deps.UserRepo.Save(ctx, &updated); err != nil {
		return nil, err
	}
	if err := deps.PasswordHistoryRepo.Add(ctx, user.Sub, encoded, now); err != nil {
		return nil, err
	}
	if deps.Emit != nil {
		deps.Emit(&spec.PasswordChanged{At: now, TenantID: user.TenantID, Sub: user.Sub})
		if clearedUpdatePassword {
			// 自動解除なので ActorSub は本人 (system 操作ではなく能動的解除)。
			deps.Emit(&spec.UserRequiredActionCleared{
				At: now, TenantID: user.TenantID, ActorSub: user.Sub, TargetSub: user.Sub,
				Action: string(spec.RequiredActionUpdatePassword),
			})
		}
	}
	return &updated, nil
}

File: ra-idp/src/authentication/usecases/change-password.ts
/**
 * Layer 3 — Application Logic（change-password use case）
 *
 * 認証済みユーザーが自身のパスワードを変更する。SCL `interfaces.ChangePassword` /
 * `events.PasswordChanged` / invariant `PasswordHistoryNoReuse` の双子。
 *
 * フロー:
 *   1. sub から User を解決
 *   2. current_password を hash と verify
 *   3. new_password を password_policy で検証
 *   4. 直近 history_depth 件と逐次 verify（一致したら拒否）
 *   5. new_password を hash → save → history.add → PasswordChanged emit
 */

import type { DomainEvent, User } from '../../spec-bindings/schemas'
import type { PasswordHasher } from '../ports/password-hasher'
import type { PasswordHistoryRepository } from '../ports/password-history-repository'
import type { UserRepository } from '../ports/user-repository'
import { PASSWORD_POLICY, PasswordPolicyError, validatePassword } from './password-policy'

export class CurrentPasswordMismatchError extends Error {
  constructor() {
    super('current password does not match')
    this.name = 'CurrentPasswordMismatchError'
  }
}

export class PasswordReuseError extends Error {
  constructor() {
    super('new password matches a recent password')
    this.name = 'PasswordReuseError'
  }
}

export class UserNotFoundError extends Error {
  constructor(public readonly sub: string) {
    super(`user not found: ${sub}`)
    this.name = 'UserNotFoundError'
  }
}

export interface ChangePasswordDeps {
  userRepo: UserRepository
  passwordHasher: PasswordHasher
  historyRepo: PasswordHistoryRepository
  emit: (e: DomainEvent) => void
  /** 将来テナント別ポリシーで上書きする際の差し込み点。省略時は global default。 */
  historyDepth?: number
}

export interface ChangePasswordInput {
  sub: string
  current_password: string
  new_password: string
  now?: Date
}

export async function changePassword(
  deps: ChangePasswordDeps,
  input: ChangePasswordInput,
): Promise<User> {
  const user = await deps.userRepo.findBySub(input.sub)
  if (!user) throw new UserNotFoundError(input.sub)

  const currentOk = await deps.passwordHasher.verify(input.current_password, user.password_hash)
  if (!currentOk) throw new CurrentPasswordMismatchError()

  const result = validatePassword(input.new_password, {
    username: user.preferred_username,
    email: user.email,
  })
  if (!result.ok) throw new PasswordPolicyError(result.violations)

  const depth = deps.historyDepth ?? PASSWORD_POLICY.historyDepth
  const recent = await deps.historyRepo.recent(user.sub, depth)
  for (const entry of recent) {
    if (await deps.passwordHasher.verify(input.new_password, entry.encoded)) {
      throw new PasswordReuseError()
    }
  }
  if (await deps.passwordHasher.verify(input.new_password, user.password_hash)) {
    throw new PasswordReuseError()
  }

  const now = input.now ?? new Date()
  const newHash = await deps.passwordHasher.hash(input.new_password)
  const updated: User = { ...user, password_hash: newHash, updated_at: now.toISOString() }
  await deps.userRepo.save(updated)
  await deps.historyRepo.add(user.sub, newHash, now)
  deps.emit({ type: 'PasswordChanged', occurredAt: now.toISOString(), sub: user.sub })
  return updated
}
